### PR TITLE
feat: add support for yarn berry lockfiles

### LINKED ISSUE #101: Yarn v2 support needed
**Is your feature request related to a problem? Please describe.**

Yarn v2 introduced a new lockfile format that now adheres to YAML. I believe ATM when trying to run lockfile-lint on a project with yarn v2, lockfile-lint tries to use the v1 parser and fails with an error.

```
ABORTING lockfile lint process due to error exceptions

Unable to parse lockfile "yarn.lock"

SyntaxError: Unknown token: {line: 3, col: 2, type: ''INVALID', value: undefined } 3:2 in lockfile
```

**Describe the solution you'd like**

As far as I can tell, yarn v2 does not have an equivalent lockfile parser. But since it adheres to YAML now, I'm wondering if it is possible to just use a parser like [yaml](https://github.com/eemeli/yaml) and create a small function to transform it like `parseNpmLockfile`?

**Describe alternatives you've considered**

None.

### PR DESCRIPTION
**Pull request by brad-decker**
## Description
Adds minimal support for yarn berry lockfiles. This is done by checking if the lockfile is yarn berry by referencing the initial key in the lockfile which will now be the __metadata key. If this occurs the parser normalizes the structure to match what is expected downstream. The changes that are here will result in users of this repo with yarn berry to change the way they configure allowed schemes (the resolved schemes are now things like 'npm:', 'patch:' and 'portal:'. All tests pass for the v1 lock file as well

Example use-case:

```
lockfile-lint --path yarn.lock --allowed-hosts npm yarn github.com codeload.github.com --empty-hostname true --allowed-schemes "https:" "git+https:" "npm:" "patch:" "workspace:"
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Related Issue

fixes #101 

## How Has This Been Tested?
A new lockfile using yarn berry was generated and added to __fixtures__ and a new test was added to load and parse that lockfile

## Checklist:

- [ ] I have updated the documentation (if required).
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] I added a picture of a cute animal cause it's fun

### CODE REVIEW

**Review by lirantal — CHANGES_REQUESTED** [review]
Small changes

File: packages/lockfile-lint-api/package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "lockfile-lint-api",
-  "version": "5.4.5",
+  "version": "5.4.6",
```
[the remark this answers is not in the document]
Comment (lirantal) [reply packages/lockfile-lint-api/package.json:3]:
```suggestion
  "version": "5.4.5",
```

File: packages/lockfile-lint/package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "lockfile-lint",
-  "version": "4.9.5",
+  "version": "4.9.6",
```
Comment (lirantal) [inline packages/lockfile-lint/package.json:3]:
```suggestion
  "version": "4.9.5",
```

**Review by lirantal — CHANGES_REQUESTED** [review]

**Comment by lirantal** [comment]
This looks good to me. I'm happy if @naugtur wants to take a pass at it too.
@brad-decker notice just really the explicit version change that you've updated which we need to revert (I committed to overwrite it so it's now ok).

Overall looks good to me. Let's give it a few days for more eyes before we merge.

File: packages/lockfile-lint-api/src/ParseLockfile.js [hunk 3]
```diff
@@ -20,14 +20,14 @@ const {
  * Checks if a sample object is a valid dependency structure
  * @return boolean
  */
-function checkSampleContent (lockfile) {
-  const [sampleKey, sampleValue] = Object.entries(lockfile)[0]
+function checkSampleContent (lockfile, isYarnBerry) {
```
Comment (naugtur) [inline packages/lockfile-lint-api/src/ParseLockfile.js:23]: new implementation may diverge further (we might discover or they might add new differences) - it'd be best if these were separate functions IMHO.  
Does anybody feel it's better to DRY it?

File: packages/lockfile-lint-api/package.json [hunk 4]
```diff
@@ -49,7 +49,7 @@
     "directory": "packages/lockfile-lint-api"
   },
   "dependencies": {
-    "@yarnpkg/parsers": "^3.0.0-rc.6",
+    "@yarnpkg/parsers": "^3.0.0-rc.32",
```
[the remark this answers is not in the document]
Comment (lirantal) [reply packages/lockfile-lint-api/package.json:52]: You're referring to the yarn repo, right?

File: packages/lockfile-lint-api/src/ParseLockfile.js (hunk 3 above)
Comment (lirantal) [reply packages/lockfile-lint-api/src/ParseLockfile.js:23]: @naugtur you mean to have a separate function to check for yarn's version (1, 2, 3, etc) ? I guess in this part specifically of parsing the lockfile we really just need to ensure it is a parsable one, so perhaps we simply take in the detected lockfile type and then call one of the package manager's own lib & function to parse, and as long as it doesn't throw, we're ok here.

I don't mind doing that in another new PR though if @brad-decker is up for it.

File: packages/lockfile-lint-api/src/ParseLockfile.js [hunk 5]
```diff
@@ -36,11 +36,30 @@ function checkSampleContent (lockfile) {
...
   const lockfile = yarnParseSyml(lockfileBuffer.toString())
+  const isYarnBerry = typeof lockfile.__metadata === 'object'
   const hasSensibleContent =
-    lockfile && typeof lockfile === 'object' && checkSampleContent(lockfile)
+    lockfile && typeof lockfile === 'object' && checkSampleContent(lockfile, isYarnBerry)
   if (!hasSensibleContent) {
     throw Error('Lockfile does not seem to contain a valid dependency list')
   }
+
+  if (isYarnBerry) {
```
Comment (lirantal) [inline packages/lockfile-lint-api/src/ParseLockfile.js:46]: realistically this part is where we're really want to branch out to specific functions, each one handling its own lockfile version (yarn1, berry, etc). Again, I don't mind merging this PR and then shooting a new PR to refactor this properly.

File: packages/lockfile-lint-api/src/ParseLockfile.js (hunk 3 above)
Comment (lirantal) [reply packages/lockfile-lint-api/src/ParseLockfile.js:23]: @naugtur you mean to have a specific function to check for yarn's version (1, 2, 3, etc)

**Comment by kachkaev** [comment]
Glad to see progress on #101! Happy to help with testing in https://github.com/blockprotocol/blockprotocol/pull/680, if needed 🚀

File: packages/lockfile-lint-api/__tests__/parseYarnLockfile.test.js [hunk 6]
```diff
@@ -24,6 +24,28 @@ describe('ParseLockfile Yarn', () => {
...
+  it('parsing a yarn berry lockfile returns an object with packages', () => {
+    const mockYarnLockfilePath = path.join(__dirname, './__fixtures__/yarnberry.lock')
+    const options = {
+      lockfilePath: mockYarnLockfilePath,
+      lockfileType: 'yarn'
+    }
+    const parser = new ParseLockfile(options)
+    const lockfile = parser.parseSync()
+
+    // console.log(lockfile)
```
Comment (lirantal) [inline packages/lockfile-lint-api/__tests__/parseYarnLockfile.test.js:36]:
```suggestion
```

remove console.log comment

File: packages/lockfile-lint/package.json [hunk 7]
```diff
@@ -55,7 +55,7 @@
   "dependencies": {
     "cosmiconfig": "^7.0.1",
     "debug": "^4.1.1",
-    "lockfile-lint-api": "^5.4.5",
+    "lockfile-lint-api": "^5.4.6",
```
Comment (lirantal) [inline packages/lockfile-lint/package.json:58]:
```suggestion
```

Comment (lirantal) [reply packages/lockfile-lint/package.json:58]: don't need to touch the version, it will be updated automatically

**Review by lirantal — COMMENTED** [review]
suggesting commit updates to land these changes

File: packages/lockfile-lint/package.json [hunk 8]
```diff
@@ -55,7 +55,7 @@
   "dependencies": {
     "cosmiconfig": "^7.0.1",
     "debug": "^4.1.1",
-    "lockfile-lint-api": "^5.4.5",
+    "lockfile-lint-api": "^5.4.5"
```
Comment (lirantal) [inline packages/lockfile-lint/package.json:58]:
```suggestion
    "lockfile-lint-api": "^5.4.5",
```

File: packages/lockfile-lint-api/__tests__/parseYarnLockfile.test.js [hunk 9]
```diff
@@ -24,6 +24,27 @@ describe('ParseLockfile Yarn', () => {
...
 
+  it('parsing a yarn berry lockfile returns an object with packages', () => {
+    const mockYarnLockfilePath = path.join(__dirname, './__fixtures__/yarnberry.lock')
+    const options = {
+      lockfilePath: mockYarnLockfilePath,
+      lockfileType: 'yarn'
+    }
+    const parser = new ParseLockfile(options)
+    const lockfile = parser.parseSync()
+
```
Comment (lirantal) [inline packages/lockfile-lint-api/__tests__/parseYarnLockfile.test.js:35]:
```suggestion
```

**Review by lirantal — COMMENTED** [review]
fixing lint errors

**Comment by lirantal** [comment]
👋🏽  @brad-decker @naugtur @kumavis  FYI that this has landed and you can apply any relevant updates on the metamask project/repo

**Comment by brad-decker** [comment]
Sorry for dropping the ball on this @lirantal but thank you for getting it landed. Much appreciated!

**Comment by lirantal** [comment]
All good buddy. Thank you for opening this to begin with :-)

**Pull request merged**